ruma-common/src/push/condition.rs

```diff
@@ -42,13 +52,273 @@ pub enum PushCondition {
+    }
+}
+
+/// Additional functions for string matching.
+pub trait StrExt {
+    /// Get the length of the char at `index` in `string`. The byte index must correspond to
+    /// the start of a char boundary.
+    fn char_len(&self, index: usize) -> usize;
+
+    /// Get the char at `index` in `string`. The byte index must correspond to the start of
+    /// a char boundary.
+    fn char_at(&self, index: usize) -> char;
+
+    /// Get the index of the char that is before the char at `index` in `string`. The byte index
+    /// must correspond to a char boundary. Returns a tuple with the index and the char.
+    fn find_prev_char(&self, index: usize) -> Option<char>;
+
+    /// Matches the given value against a pattern.
+    ///
+    /// The match is case insensitive.
+    ///
+    /// If `match_words` is true, looks for pattern as a substring of `value`,
+    /// and checks that it is a "word". If it is false, checks the pattern as
+    /// a glob with wildcards `*` and `?`.
+    fn matches_pattern(&self, pattern: &str, match_words: bool) -> bool;
+
+    /// Matches the given value against a pattern, with word boundaries.
+    ///
+    /// The match is case sensitive.
+    fn matches_word(&self, pattern: &str) -> bool;
+}
+
+impl StrExt for str {
+    fn char_len(&self, index: usize) -> usize {
+        let mut len = 1;
+        while !self.is_char_boundary(index + len) {
+            len += 1;
+        }
+        len
+    }
+
+    fn char_at(&self, index: usize) -> char {
+        let end = index + self.char_len(index);
+        let char_str = &self[index..end];
+        char::from_str(char_str)
+            .unwrap_or_else(|_| panic!("Could not convert str '{}' to char", char_str))
+    }
+
+    fn find_prev_char(&self, index: usize) -> Option<char> {
+        if index == 0 {
+            return None;
+        }
+
+        let mut pos = index - 1;
+        while !self.is_char_boundary(pos) {
+            pos -= 1;
+        }
+        Some(self.char_at(pos))
+    }
+
+    fn matches_pattern(&self, pattern: &str, match_words: bool) -> bool {
+        if self.is_empty() || pattern.is_empty() {
+            return false;
+        }
+
+        let value = &self.to_lowercase();
+        let pattern = &pattern.to_lowercase();
+
+        if match_words {
+            value.matches_word(pattern)
+        } else {
+            WildMatch::new(pattern).matches(value)
+        }
+    }
+
+    fn matches_word(&self, pattern: &str) -> bool {
+        match self.find(pattern) {
+            Some(start) => {
+                let end = start + pattern.len();
+
+                // Look if the match has word boundaries.
+                let word_boundary_start = if !self.char_at(start).is_word_char() {
+                    true
+                } else {
+                    match self.find_prev_char(start) {
+                        Some(c) => !c.is_word_char(),
+                        None => true,
+                    }
+                };
```

Comment:
This *could* be shortened to

```suggestion
                let word_boundary_start = !self.char_at(start).is_word_char() ||
                    match self.find_prev_char(start) {
                        Some(c) => !c.is_word_char(),
                        None => true,
                    };
```

or, replacing the `match` with combinators

```suggestion
                let word_boundary_start = !self.char_at(start).is_word_char() ||
                    self.find_prev_char(start).map_or(true, |c| !is_word_char(c));
```

But I'm not sure how `rustfmt` will format any of this. I'm fine keeping this as-is if you prefer.

Reply:
I'll go with your second suggestion